### DISCUSSION TITLE
VReplication: Where does the param of VStreamerClient from?

### DISCUSSION THREAD

**Discussion by earayu** [wesql/wescale#301] [Q&A] [answered]
VStramer is a service that anyone can call, as long as they provide the correct params.
But where does these params come from?
But can I build right params so that I can call VStreamerClient?
![image](https://github.com/wesql/wescale/assets/13747699/75bb309d-bb45-4635-b8b5-8472133b9bb8)

**Comment by gerayking** [comment, marked as answer]
The params for Vstream is sourced from `vplayer`, as can be seen in the `fetchAndApply` function. 
The params for VstreamRows is sourced from `vcopier,` as can be seen in the `copyTable` function.

**Reply by earayu** [reply]
fundamentally, `startPos` in `VStream` comes from system table `vreplication.pos`

**Reply by earayu** [reply]
`tabkePKs` in VStream is nil

**Reply by earayu** [reply]
`filter` in `VStream` is from system table `vreplication.source`.
And so dose for `query` in `VStreamRows`

**Reply by earayu** [reply]
`lastpk` in `VStreamRows` is from system table `copy_state.lastpk`
